### ISSUE TITLE
Segfault on push of undefined name

### ISSUE THREAD

**Issue by kevincox** [hasithvm/lasm#26]
When compiling the following file I get a segfault with no error message.  Tested on 89ecc3fbb4ea3b8b4a07941f7f2f74ce89f4df9a and tag 1.2.2

``` asm
; Lab6-P1 - Subroutines to print integers, digits, and salaries

; Constant definitions
Display .EQU 04E9h  ; address of Libra display


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;  START OF SUBROUTINES to COPY to lab6-P2.asm
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;;;;;;;;;;;;;;;;;
; printInt: Subroutine to print a 1-byte unsigned (short) integer between 0-255
; Input parameters:
;   Stack: Unsigned short int to be printed
; Output parameters:
;   None.
printInt:
    ; Save registers modified by this subroutine
    PUSH AX
    PUSH CX
    PUSH BI

    ; Retrieve input parameter from stack into AL
    mov BI, SP      ; FIX ME - make a copy of SP
    mov AX, [BI+8]  ; FIX ME - read from the appropriate slot of your stack frame

    mov CL, 10      ; Will be dividing by 10...

LoopPI:
    CMP AL, 10      ; Compare the number to 10
    JL printLast    ; If it is less than 10, then print this digit
    ; If it is greater than 10, divide by 10
    mov AH, 0       ; Clear AH
    DIV CL          ; Divide number by 10
    push AL         ; FIX ME - Push the input parameter of printDigit onto the stack
    call printDigit ; Print the quotient in AL
    pop AL          ; FIX ME - Pop the input parameter of printDigit back off the stack
    mov AL, AH      ; Move remainder into AL to be printed
    jmp LoopPI      ; Jump back to the top of the loop
printLast:
    push AL         ; FIX ME - Push the input parameter of printDigit onto the stack
    call printDigit ; Print the quotient in AL
    pop AL          ; FIX ME - Pop the input parameter of printDigit back off the stack

    ; Restore registers
    pop BI
    pop CX
    pop AX

    RET


;;;;;;;;;;;;;;;;;
; printDigit: Subroutine to print a single decimal digit
; Input parameters:
;   Stack: Unsigned decimal digit (between 0-9) to be printed
; Output parameters:
;   None.
printDigit:
    ; Save registers modified by this subroutine
    PUSH AX
    PUSH DX
    PUSH BI         ; FIX ME

    ; Retrieve input parameter from stack into AL
    mov BI, SP      ; FIX ME - make a copy of SP
    mov AX, [SP+8]  ; FIX ME - read from the appropriate slot of your stack frame

    mov DX, Display
    ADD AL, '0'     ; Convert number to ASCII code
    OUT DX,AL       ; Print it

    ; Restore registers
    pop BI          ; FIX ME
    pop DX
    pop AX

    RET


;;;;;;;;;;;;;;;;;
; printSalary: Subroutine to print employee salary
; Input parameters:
;   Stack: Unsigned short int (0-255) representing salary in 1000's of $
; Output parameters:
;   None.

printSalary:
    ; Save registers modified by this subroutine
    PUSH AX
    PUSH SI             ; Not strictly necessary, but please keep
    PUSH DX
    PUSH BI             ; FIX ME

    ; Retrieve input parameter from stack into AL
    mov BI, SP           ; FIX ME - make a copy of SP
    mov AX, [BI+8]       ; FIX ME - read from the appropriate slot of your stack frame

    mov DX, Display

    mov AH,AL           ; Keep a copy of the salary in AH (need AL for printing...)
    mov AL, '$'         ; Print '$' preceeding number
    OUT DX,AL           ; Print it
    mov AL,AH           ; Move salary back into AL
    push AX             ; FIX ME - Push the input parameter of printInt onto the stack
    call printInt       ; Print the salary (0-255)
    push AX             ; FIX ME - Pop the input parameter of printInt back off the stack

    mov AL, ','         ; Print ',' after number
    OUT DX,AL           ; Print it
    mov AL, '0'         ; Print '0' after comma
    OUT DX,AL           ; Print a zero
    OUT DX,AL           ; Print a zero
    OUT DX,AL           ; Print a zero

    ; Restore registers
    pop BI            ; FIX ME
    pop DX
    pop SI
    pop AX

    RET


;;;;;;;;;;;;;;;;;
; newLine: Subroutine to print a newline and a linefeed character
; Input parameters:
;   None.
; Output parameters:
;   None.

; Constants for this subroutine:
s_CR .EQU 0Dh       ; ASCII value for Carriage return
s_LF .EQU 0Ah       ; ASCII value for NewLine

newLine:
    ; Save registers modified by this subroutine
    PUSH AX
    PUSH DX

    mov DX, Display     ; Initialize the output port number in DX
    mov AL, s_LF        ; Load line feed (LF) into AL
    out DX,AL           ; print the char
    mov AL, s_CR        ; Load carriage return (CR) into AL
    out DX,AL           ; print the char

    ; Restore registers
    pop DX
    pop AX

    RET

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
; END OF SUBROUTINES FOR lab6-P2.asm
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
; Test data
;

.ORG 00A0h

num1: .DB 86                    ; Should print as decimal 86
sal1: .DB 34                ; Should print as '$34,000'

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;    
; main: Main function to test all subroutines
.ORG 00B0h

main:

    ; Print a short unsigned int (0-99). Use num1
    mov AL, [num1]              ; FIX ME - get input parameter
    push AX                     ; FIX ME - place input parameter on stack
    call printInt
    pop AX                      ; FIX ME - remove input parameter from stack
    call newLine

    ; Print a salary. Use sal1
    mov AL, [sal1]              ; FIX ME - get input parameter
    push AX                     ; FIX ME - place input parameter on stack
    call printSalary
    pop AX                      ; FIX ME - remove input parameter from stack
    call newLine

    ;Quit
    HLT



.END main       ;Entry point of program is main()
```

**Comment by kevincox** [comment]
It appears that mentioning BI causes the to crash.  (BP is of course the base pointer but it probably shouldn't segfault)

**Comment by kevincox** [comment]
I have managed to simplify the test case.

``` asm
main:
    push undef
    hlt

.END main
```

**Comment by hasithvm** [comment]
This issue was closed with 973687bd8d8c529fe3416569c524650c030337b6.
